### PR TITLE
accept non-string data as input (lists or tuples)

### PR DESCRIPTION
**Pull request by internaut**
This patch allows to use the DL distance also for arbitrary non-string data, e.g.:

    >>> damerau_levenshtein_distance([1, 3, 2], [1, 2, 3])
    1

Lists and tuples of arbitrary data can be passed. Please note that NumPy arrays have to be converted to lists before using them as inputs.

Also updated tests for non-string data.

Additionally fixed minor issue with TypeError exception message in `_to_unicode`.

### CODE REVIEW

**Comment by gfairchild** [comment]
This is cool. To allow this to pass the tests, I'm fine with dropping tests for Python 3.2 and 3.3. Would you mind removing those tests in `.travis.yml`?

**Comment by internaut** [comment]
Python 3.3 didn't fail in the tests, only 3.2 and it looks like it's only because I created the c-file with a newer version of Cython that dropped support for 3.2: `#error Cython requires Python 2.6+ or Python 3.3+`
Should I drop 3.2 from Travis anyway or do you want to "cythonize" with an older version of Cython?

**Comment by gfairchild** [comment]
Oh, sorry, I should've been clearer. I meant to get rid of 3.2 and 3.3 because they're past their end of life dates. I don't really care to support unsupported versions of Python. Let's go ahead and drop both. The readme should also be updated to state which versions we're officially supporting.

File: pyxdameraulevenshtein/pyxdameraulevenshtein.pyx [hunk 1]
```diff
@@ -45,7 +45,7 @@ cdef unicode _to_unicode(s):
     elif isinstance(s, unicode):
         # this works for NumPy strings
         return unicode(s)
-    raise TypeError('string [{}] has an unrecognized type of [{}]'.format(type(s)))
+    raise TypeError('string [{}] has an unrecognized type of [{}]'.format(s, type(s)))
```
Comment (gfairchild) [inline pyxdameraulevenshtein/pyxdameraulevenshtein.pyx]: Nice catch.

File: pyxdameraulevenshtein/pyxdameraulevenshtein.pyx [hunk 2]
```diff
@@ -63,8 +63,12 @@ cpdef unsigned long damerau_levenshtein_distance(seq1, seq2):
         >>> damerau_levenshtein_distance('orange', 'pumpkin')
         7
     """
-    s1 = _to_unicode(seq1)
-    s2 = _to_unicode(seq2)
+    if (isinstance(seq1, list) or isinstance(seq1, tuple)) and (isinstance(seq2, list) or isinstance(seq2, tuple)):
```
Comment (gfairchild) [inline pyxdameraulevenshtein/pyxdameraulevenshtein.pyx]: We can simplify this line a bit to:

```python
if isinstance(seq1, (list, tuple)) and isinstance(seq2, (list, tuple)):
```

File: pyxdameraulevenshtein/pyxdameraulevenshtein.pyx [hunk 3]
```diff
@@ -137,7 +141,12 @@ cpdef float normalized_damerau_levenshtein_distance(seq1, seq2):
         >>> normalized_damerau_levenshtein_distance('orange', 'pumpkin')
         1.0
     """
-    return float(damerau_levenshtein_distance(seq1, seq2)) / max(len(_to_unicode(seq1)), len(_to_unicode(seq2)))
+    if (isinstance(seq1, list) or isinstance(seq1, tuple)) and (isinstance(seq2, list) or isinstance(seq2, tuple)):
```
Comment (gfairchild) [inline pyxdameraulevenshtein/pyxdameraulevenshtein.pyx]: We can make the same change here as proposed on line `66`.

File: tests/test_pyxdl.py [hunk 4]
```diff
@@ -21,6 +21,9 @@ def test_damerau_levenshtein_distance(self):
         assert damerau_levenshtein_distance('orange', 'pumpkin') == 7
         assert damerau_levenshtein_distance('gifts', 'profit') == 5
         assert damerau_levenshtein_distance('Sjöstedt', 'Sjostedt') == 1
+        assert damerau_levenshtein_distance([1, 2, 3], [1, 3, 2]) == 1
+        assert damerau_levenshtein_distance((1, 2, 3), (1, 3, 2)) == 1
+        assert damerau_levenshtein_distance((1, 2, 3), [1, 3, 2]) == 1
```
Comment (gfairchild) [inline tests/test_pyxdl.py]: Could we possibly add a couple more slightly more complex tests here? For example, test tuples and also test for a couple non-1 results.

File: .travis.yml [hunk 5]
```diff
@@ -3,8 +3,6 @@ language: python
 python:
   - 2.6
```
Comment (gfairchild) [inline .travis.yml]: And actually, let's remove 2.6 support too. 2.6 is end of life as well. We may support 2.6 or 3.3 by happenstance, but it's not something I want to strive to maintain.

File: README.md [hunk 6]
```diff
@@ -13,7 +13,7 @@ pyxDamerauLevenshtein implements the Damerau-Levenshtein (DL) edit distance algo
 This implementation is based on [Michael Homer's pure Python implementation](https://web.archive.org/web/20150909134357/http://mwh.geek.nz:80/2009/04/26/python-damerau-levenshtein-distance/), which implements the [optimal string alignment distance algorithm](https://en.wikipedia.org/wiki/Damerau%E2%80%93Levenshtein_distance#Optimal_string_alignment_distance). It runs in `O(N*M)` time using `O(M)` space. It supports unicode characters.
 
 ## REQUIREMENTS
-This code requires Python 2.6+ (including Python 3.2+), [NumPy](http://www.numpy.org/), and a C compiler such as GCC. Although the code was written in Cython, Cython is not required for installation.
+This code requires Python 2.6, 2.7 or Python 3.4+, [NumPy](http://www.numpy.org/), and a C compiler such as GCC. Although the code was written in Cython, Cython is not required for installation.
```
Comment (gfairchild) [inline README.md]: Here, let's remove the 2.6 mention.

File: .travis.yml (hunk 5 above)
Comment (gfairchild) [reply .travis.yml]: Additionally, let's remove `setup.py:80,83-84` to indicate 2.6 and 3.2-3.3 removal.

**Comment by gfairchild** [comment]
And finally, one last suggestion: would you mind modifying `CHANGES.md` to indicate your changes? We can call this release 1.5. Also, edit `setup.py:44` to state version 1.5.

**Comment by internaut** [comment]
I came across a division-by-zero in `normalized_damerau...` when comparing empty input strings / lists.
I also added some more tests (also to catch this error) but for better testing you should integrate something like [hypothesis](http://hypothesis.works/). But that's something for the future, maybe.
Regarding Python version support, I dared to add 3.6 since this is the most recent version.

**Comment by gfairchild** [comment]
This is great. Accepting. Thank you!

**Comment by internaut** [comment]
Cool, thanks! Will you upload a new version to PyPI? Otherwise the README/documentation on the repo does not reflect the state in PyPI.

**Comment by gfairchild** [comment]
Yeah, I'll prioritize the updated PyPI release in the next day or two. Good point.

**Comment by gfairchild** [comment]
It's live on PyPI! Thanks again for your contributions. I attributed you in `AUTHORS.md` and `CHANGES.md`.

**Comment by internaut** [comment]
Cool, thanks! Was a pleasure to collaborate!

**Pull request merged**